src/validation.cpp

```diff
@@ -465,6 +464,111 @@ static bool CheckInputsFromMempoolAndCache(const CTransaction& tx, TxValidationS
     return CheckInputScripts(tx, state, view, flags, /* cacheSigStore = */ true, /* cacheFullSciptStore = */ true, txdata);
 }
 
+/**
+ * A CoinsView that adds a memory cache to another CoinsView and serves as temporary scratch space.
+ * Used by MemPoolAccept class to validate transactions and packages before submitting to mempool.
+ * A backend can be set to provide read access to chainstate and/or mempool coins, but writing to
+ * the backend is disabled. Avoid using a CoinsViewTemporary in consensus-critical paths such
+ * as writing to the script cache. See CheckInputsFromMempoolAndCache as an example. When not being
+ * used to validate a package (m_temp_added and m_temp_spent are empty), a CoinsViewTemporary
+ * behaves exactly like a CCoinsViewCache.
+ */
+class CoinsViewTemporary : public CCoinsViewCache
+{
+    /**
+    * Coins made available by transactions being validated. Tracking these allows for package
+    * validation, since we can access transaction outputs without submitting them to mempool.
+    */
+    std::map<COutPoint, Coin> m_temp_added;
+
+    /**
+    * Coins spent by transactions being validated. When validating a package, we need to track
+    * these in order to distinguish between missing/spent coins and conflicts within a package.
+    */
+    std::set<COutPoint> m_temp_spent;
+
+    /** An empty coin used as a placeholder for a spent coin.*/
+    inline static const Coin s_coin_spent;
```

Comment:
nit: Hmmm... I know GetCoin is perf sensitive but do we do this placeholder coin often?

Would optional coin be cleaner?

Reply:
We had a conversation about the Coins API being a little weird earlier: https://github.com/bitcoin/bitcoin/pull/20833#discussion_r586764973

In any case, it's out of the scope of this PR to change the Coins API, so I'm going to leave it as is :shrug: